big-o/big-o-analysis.md

```diff
@@ -0,0 +1,335 @@
+   1. All 6 of the operations inside the `while` loop can happen _n_ times
+
+4. Create a formula that counts how many operations there are:
+
+   1. We can express this as _4 + (6 \* n)_, or _4 + 6n_
+
+5. Drop the constants:
+
+   1. We are able to drop `4 +` and `6 *`
+   1. After dropping the constants, our remaining formula is _n_
+
+6. Match this formula with the most relevant complexity:
+   1. _n_ matches _O(n)_ the most closely. This algorithm has linear time complexity.
+
+### What is "Dropping the Constant," or Finding the Dominant Order?
+
+In complexity analysis, programmers care about the dominant order, or the part of the complexity that affect the complexity _the most._
+
+Mathematical constants are values that have a fixed value. Constants tend to be added, subtracted, multiplied, or divided in our formulas. In Big O, constants will affect our complexity (O(6n) is different than O(n))... but they do not affect the curve enough that is meaningful for complexity analysis.
+
+Some examples:
+
+- O(2n) is shortened to O(n)
+- O(n/2) (or one half of _n_) is shortened to O(n)
+- O(n + n<sup>2</sup>) is shortened to O(n<sup>2</sup>)
+
+## Space Complexity: Code to Big O
+
+Recall that space complexity is the measurement of how much memory an algorithm uses as the size of the input changes. Space complexity is measured in the _amount of memory allocated in this algorithm._ This amount of memory doesn't include the memory that stores our data set. Our goal will be to count the number of variables needed.
+
+### !callout-info
+
+## Lists Require More Memory Than A Single Value
+
+We can imagine that a single value `x = 42` will take "one unit" of memory, because it is a variable that holds one value. Following that logic, a list with many values, such as `[42, 43, 44, 45]` will take "four units" of memory. The amount of memory that the list holds is _n_ amount, in terms of space complexity.
+
+We will get into the details of lists and memory in a future lesson.
+
+### !end-callout
+
+### Generic Steps
+
+1. Read through the code, and identify all places where a variable is _initially assigned._
+   - We do not need to count re-assignment of variables, as the value will occupy the same memory as the initial assignment
+1. Recognize which variables have a value that could take a variable amount of memory
+   - Typically, storing a list will require _n_ amount of memory
+1. Create an equation that represents how many values are initialized and stored in memory
+1. Drop the constants
+1. Match this Big O to a relevant complexity
+
+### Small Example
+
+Given an implementation of linear search...
+
+```python
+def linear_search(array, item):
+    for i in range(len(array)):
+        if item == array[i]:
+            return i
+    return False
+```
+
+1. Identify all places where the algorithm initializes a variable. (Don't count the input data set!)
+
+   - `i`
+   - `item`
```

Comment:
if the input data set (`array`, here) isn't counted then `item` probably shouldn't be either.

Reply:
Good call! I'll include both with a parenthetical, since folks will naturally question its presence if we don't include it.

```suggestion
   - `item`
   - `array`

Note: For this exercise and the purpose of illustrating space complexity, we'll consider `item` and won't consider `array`. However, it is a completely valid interpretation to not consider `item` or `array`, as they are "inputs" into this function.
```